simulator/mips/mips_instr_decode.h

```diff
@@ -1,95 +1,42 @@
- /* mips_instr_decode.h - instruction decoder for mips
+/**
+ * mips_instr_decode.h - instruction decoder for mips
  * @author Andrey Agrachev agrachev.af@phystech.edu
  * Copyright 2018 MIPT-MIPS
  */
 
 #ifndef MIPS_INSTR_DECODE_H
 #define MIPS_INSTR_DECODE_H
 
-enum instr_size_and_shifts {
-    FUNCT_SIZE    = 6,
-    SHAMT_SIZE    = 5,
-    RD_SIZE       = 5,
-    RT_SIZE       = 5,
-    RS_SIZE       = 5,
-    OPCODE_SIZE   = 6,
-    ASI_IMM_SIZE  = 16,
-    ASJ_IMM_SIZE  = 26,
-
-    FUNCT_SHIFT   = 0,
-    SHAMT_SHIFT   = FUNCT_SHIFT + FUNCT_SIZE,
-    RD_SHIFT      = SHAMT_SHIFT + SHAMT_SIZE,
-    RT_SHIFT      = RD_SHIFT    + RD_SIZE,
-    RS_SHIFT      = RT_SHIFT    + RT_SIZE,
-    OPCODE_SHIFT  = RS_SHIFT    + RS_SIZE,
-    ASI_IMM_SHIFT = 0,
-    ASJ_IMM_SHIFT = 0,
+#include <infra/types.h>
+#include <infra/macro.h>
+
+struct MIPSInstrDecoder
+{
+    const uint32 funct;
+    const uint32 shamt;
+    const uint32 rd;
+    const uint32 rt;
+    const uint32 rs;
+    const uint32 opcode;
+    const uint32 imm;
+    const uint32 jump;
+
+    static constexpr uint32 apply_mask(uint32 bytes, uint32 mask) noexcept
```

Comment:
This is a much better solution

Reply:
Basically it's the same code as yours, just refactored a little.